gcc/rust/resolve/rust-resolve-builtins.cc

```diff
@@ -0,0 +1,107 @@
+// Copyright (C) 2025 Free Software Foundation, Inc.
+
+// This file is part of GCC.
+
+// GCC is free software; you can redistribute it and/or modify it under
+// the terms of the GNU General Public License as published by the Free
+// Software Foundation; either version 3, or (at your option) any later
+// version.
+
+// GCC is distributed in the hope that it will be useful, but WITHOUT ANY
+// WARRANTY; without even the implied warranty of MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+// for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with GCC; see the file COPYING3.  If not see
+// <http://www.gnu.org/licenses/>.
+
+#include "rust-resolve-builtins.h"
+#include "rust-name-resolution-context.h"
+#include "rust-tyty.h"
+#include "rust-hir-type-check.h"
+
+namespace Rust {
+namespace Resolver2_0 {
+namespace Builtins {
+
+class TypeBuilder
+{
+public:
+  constexpr TypeBuilder (const char *name, TyTy::BaseType *(*type_gen) (HirId))
+    : name (name), type_gen (type_gen)
+  {}
+
+  std::string create_name () const { return name; }
+
+  TyTy::BaseType *create_type (HirId hir_id) const { return type_gen (hir_id); }
+
+private:
+  const char *name;
+  TyTy::BaseType *(*type_gen) (HirId);
+};
+
+#define GEN_DEF(name, expr)                                                    \
+  TypeBuilder ((name), [] (HirId hir_id) -> TyTy::BaseType * { return (expr); })
```

Comment:
```suggestion
  TypeBuilder ((name), [] (HirId hir_id) -> TyTy::BaseType * { return new (expr); })
```

?

feel free to ignore